crates/sui-graphql-rpc/src/types/transaction_block_effects.rs

```diff
@@ -46,79 +54,111 @@ impl TransactionBlockEffects {
+            Either::Left(ref native) => native.dependencies(),
+            Either::Right(ref pre_commit_effects) => pre_commit_effects.dependencies(),
+        };
         ctx.data_unchecked::<PgManager>()
-            .fetch_txs_by_digests(self.native.dependencies())
+            .fetch_txs_by_digests(dependencies)
             .await
             .extend()
     }
 
     /// Effects to the gas object.
     async fn gas_effects(&self) -> Option<GasEffects> {
-        Some(GasEffects::from(&self.native))
+        let gas_eff = match self.effects {
+            Either::Left(ref native) => GasEffects::from(native),
+            Either::Right(ref pre_commit_effects) => {
+                GasEffects::from_json_rpc_effects(pre_commit_effects)
+            }
+        };
+        Some(gas_eff)
     }
 
     /// Shared objects that are referenced by but not changed by this transaction.
     async fn unchanged_shared_objects(&self) -> Option<Vec<UnchangedSharedObject>> {
-        Some(
-            self.native
+        Some(match self.effects {
+            Either::Left(ref native) => native
                 .input_shared_objects()
-                .into_iter()
-                .filter_map(|input| UnchangedSharedObject::try_from(input).ok())
+                .iter()
+                .filter_map(|input| UnchangedSharedObject::try_from(input.clone()).ok())
                 .collect(),
-        )
+            Either::Right(ref pre_commit_effects) => {
+                let modified: HashSet<_> = pre_commit_effects
+                    .modified_at_versions()
+                    .iter()
+                    .map(|(r, _)| *r)
+                    .collect();
+                pre_commit_effects
+                    .shared_objects()
+                    .iter()
+                    .map(|r| {
+                        if modified.contains(&r.object_id) {
+                            InputSharedObject::Mutate(r.to_object_ref())
+                        } else {
+                            InputSharedObject::ReadOnly(r.to_object_ref())
+                        }
+                    })
+                    .filter_map(|input| UnchangedSharedObject::try_from(input).ok())
+                    .collect()
+            }
+        })
     }
 
     /// The effect this transaction had on objects on-chain.
     async fn object_changes(&self) -> Option<Vec<ObjectChange>> {
-        Some(
-            self.native
-                .object_changes()
-                .into_iter()
-                .map(|native| ObjectChange { native })
-                .collect(),
-        )
+        match self.effects {
+            Either::Left(ref native) => Some(
+                native
+                    .object_changes()
+                    .into_iter()
+                    .map(|native| ObjectChange { native })
+                    .collect(),
+            ),
+
+            // TODO: implement this for pre-commit effects
+            Either::Right(ref _pre_commit_effects) => None,
```

Comment:
Decided to pause here because this is quite a large function when [implemented](https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/effects/effects_v1.rs#L161) for natives.
If all seems well in this PR, will add it as a follow up